docs/rationale.md

```diff
@@ -0,0 +1,144 @@
+tests have evolved alongside the needs of its consumers. When conventions
+change, old tests are typically updated to accommodate the new practices. That
+doesn't always happen, though, and one doesn't have to look very far to find
+examples of tests which contradict the preferred patterns.
+
+For instance:
+
+- tests which expression expectations with `throw` statements inside of
+  conditional statements rather than the assertion API implemented by the
+  harness files (though this explicitness will always been desirable when
+  asserting the semantics of conditional statements and `throw` statements
+  themselves)
+- tests with file names derived from section numbers in the 5th edition of
+  ECMA262, e.g. `built-ins/Array/15.4.5-1.js`
+- tests which validate multiple behaviors, using elaborate comment blocks to
+  designate sections
+- tests which use deprecated harness functions, e.g. `verifyEnumerable`,
+  `verifyConfigurable`, and `verifyWritable`
+
+Since existing tests do not necessarily reflect the project's current
+best-practices, it's especially important for test authors to familiarize
+themselves with the contribution guidelines.
+
+## Test generation
+
+The project includes a software tool for generating test material from abstract
+templates. The generator was designed to promote uniformity of coverage,
+particularly for parts of the grammar that are used in many productions
+(specifically: the destructuring assignment patterns introduced in the 6th
+edition of ECMA262).
+
+This tool makes it easy to introduce enormous numbers of tests. Introducing
+more tests is not a goal unto itself, though. Test262 prioritizes coherence in
+its coverage of the specification, and it recognizes that the value of tests,
+as measured by their likelihood to identify defects, varies.
+
+For these reasons, the maintainers urge restraint in the application of the
+test generation tool.
+
+## File structure
+
+For practical reasons, tests are organized in a tree structure according to the
+conventions of modern file systems. Unfortunately, this structure is not
+expressive enough to model the semantics of a rich and evolving programming
+language like ECMAScript. This means the common and crucial task of coverage
+assessment will likely always be a challenge, but strong conventions around
+file organization can help.
+
+Tests for syntax-derived operations are organized according to the language
+grammar, with directories used to describe non-terminals. Tests for built-in
+APIs are organized according to the identifiers by which they can be accessed,
+with directories used to describe the sequence of properties that can be used
+from the global scope.
+
+Directories are not generally applied beyond these limits; further
+differentiation is instead achieved through structured file names which follow
+ad-hoc conventions. This organization balances the need to group tests
+logically with the need to discover tests.
```

Comment:
I would love to have examples here as well, even just as links to a few directories that you think serve as good examples of various ad-hoc conventions.

I have also gotten the impression that some files are traditional (`name`, `constructor`, `invoked-as-function`) and some approaches are as well: for instance, separating `throws` from successful invocations. I know these might seem implied or obvious from other sentences here, but listing them out would be helpful, even if in another details file so as not to mix them with high-level.

Reply:
> I would love to have examples here as well, even just as links to a few directories that you think serve as good examples of various ad-hoc conventions.

Sounds good. I've recommended the tests for template literals as an example where tests are organized by a structured file naming scheme.

> I have also gotten the impression that some files are traditional (`name`, `constructor`, `invoked-as-function`) and some approaches are as well: for instance, separating `throws` from successful invocations. I know these might seem implied or obvious from other sentences here, but listing them out would be helpful, even if in another details file so as not to mix them with high-level.

Speaking personally, I don't hold any general expectations at that granularity. Although I might frequently use some names (or even suggest them to others), I don't know if I could formalize rules for which files are appropriate for any given language feature.